### PR TITLE
BUG: Subset slicer on Styler failed on MultiIndex with slice(None) 

### LINKED ISSUE #33562: BUG: df.style.apply(subset=...) argument only works for explicit slices but not ":"
Related to #25858 (if not the same?), reproduced on 0.25.3 and current version 1.0.3.

It seems that the `subset` argument for pd.DataFrame.style only works for slices with explicit labels, e.g. `['A',2]` but not for select-all labels, e.g. `[:,2]`. See MWE below

```python
import pandas as pd
pd.__version__ # 1.0.3

# Taken from 
# https://pandas.pydata.org/pandas-docs/stable/user_guide/style.html
def color_negative_red(val):
    color = 'red' if val < 0 else 'black'
    return 'color: %s' % color

df = pd.DataFrame([
    [-1, 3, 6],
    [-2,-3, 2]
])
df.index = pd.MultiIndex.from_tuples([('A',1),('A',2)])

# Style works for applymap
df.style.applymap(color_negative_red)

# Works for explicit slice
slicer = pd.IndexSlice[pd.IndexSlice['A',2],:]
df.loc[slicer] # valid .loc slicer
df.style.applymap(color_negative_red, subset=slicer) # works

# Breaks for select-all slice
slicer = pd.IndexSlice[pd.IndexSlice[:,2],:]
df.loc[slicer] # valid .loc slicer
df.style.applymap(color_negative_red, subset=slicer)
# TypeError: unhashable type: 'slice'
```

### PR DESCRIPTION
**Pull request by attack68**
- [x] closes #33562
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

The problem with above issue was with the function `non_reducing_slice`. It didn't correctly re-format slicing tuples on multiindices when rows or cols also contained a `slice(None)`. An additional check is now done. This method is only used for `Styler` currently.

Tests Added / reformmatted.

### CODE REVIEW

**Comment by attack68** [comment]
ready for review..

**Comment by jreback** [comment]
can you merge master

**Comment by jreback** [comment]
cc @phofl @jbrockmendel if you'd have a look

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -2407,9 +2407,14 @@ def pred(part) -> bool:
         """
         # true when slice does *not* reduce, False when part is a tuple,
         # i.e. MultiIndex slice
-        return (isinstance(part, slice) or is_list_like(part)) and not isinstance(
-            part, tuple
-        )
+        if isinstance(part, tuple):
+            # check for sub-slice:
```
Comment (jbrockmendel) [inline pandas/core/indexing.py:2411]: can you add "GH#39421" here

instead of a for loop can you do this with an `any(...)`

**Comment by phofl** [comment]
lgtm

**Comment by jreback** [comment]
thanks @attack68

**Pull request merged**